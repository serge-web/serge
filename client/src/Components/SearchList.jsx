import { Component } from 'react'
import { connect } from 'react-redux'
import { formatShortDate } from 'src/Helpers'
import classNames from 'classnames'
import { PropTypes } from 'prop-types'

import 'src/themes/App.scss'

class SearchList extends Component {
  setSelected = (item) => {
    this.props.setSelected(item)
  }

  render () {
    const that = this

    return (
      <div className='searchlist'>
        <input type='text' className='searchlist-input' key='search-templates' placeholder={this.props.placeholder} onChange={ this.props.filter } value={this.props.searchQuery} />
        <div className='searchlist-list'>
          { this.props.listData.map(function (item) {
            // onClick handlers should not contain instantiating () to pass a specific value, .bind can be used here this is a simpler
            // ES5 way to pass properties to click handlers from arrays, another option is to build a sub-component but I don't like
            // how the data moves back and forth, it breaks Reacts initial idea of 1 way data flow.

            const active = item._id === that.props.selected ? 'active' : null

            // following line includes a workaround, for when a message type doesn't include
            // a title field
            const title = item.details.title ? item.details.title : '[Title missing]'

            const date = formatShortDate(item.lastUpdated)
            return <span className={classNames({ 'searchlist-title': true, active })} href='#' onClick={that.setSelected.bind(that, item)} key={item._id}>{title} - {date}</span>
          })
          }
        </div>
      </div>
    )
  }
}

// empty mapStateToProps is here for react-redux to wire up the dispatch function to props so firing actions is possible.
const mapStateToProps = ({ currentViewURI, umpireMenu }) => ({
  currentViewURI,
  umpireMenu
})

SearchList.propTypes = {
  listData: PropTypes.array.isRequired,
  filter: PropTypes.func.isRequired,
  selected: PropTypes.string.isRequired,
  setSelected: PropTypes.func.isRequired,
  searchQuery: PropTypes.string.isRequired,
  placeholder: PropTypes.string
}

export default connect(mapStateToProps)(SearchList)
