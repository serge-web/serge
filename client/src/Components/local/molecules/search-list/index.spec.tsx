/* global it expect */
import React from 'react'
import renderer from 'react-test-renderer'
import SearchList from './index'

describe('SearchList component:', () => {
  it('renders correctly', () => {
    const listData = [{
      completed: false,
      details: {
        title: 'State of World (laydown 2)'
      },
      lastUpdated: '2019-09-30T12:37:26.705Z',
      title: 'State of World L',
      _id: 'k16eedkp',
      _rev: '1-612d7dc5d10fc81bc7459b2801c66816'
    }, {
      lastUpdated: '2019-09-30T12:37:26.705Z',
      title: 'Daily intentions',
      details: {
        title: 'Daily Intent'
      },
      completed: false,
      _id: 'k16eedkn',
      _rev: '1-cc8e8cdb01447959c266761066448382'
    }, {
      lastUpdated: '2019-09-30T12:37:26.705Z',
      title: 'Link',
      details: {
        title: 'Link'
      },
      completed: false,
      _id: 'k16eedkm',
      _rev: '1-7fa1e6dd6b4ac5b6afc45b596ee7af61'
    }, {
      lastUpdated: '2019-09-30T12:37:26.705Z',
      title: 'Chat',
      details: {
        title: 'Chat'
      },
      completed: false,
      _id: 'k16eedkl',
      _rev: '1-09ab7a18ff677cec5d9a56f02a45788d'
    }]
    const tree = renderer
      .create(
        <SearchList
          placeholder="Select template"
          listData={listData}
          setSelected={(item): void => window.alert(item)}
          activeRow={(): boolean => true}
          rowLabel={(): string => 'Label'}
          rowFilter={(): boolean => true}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
