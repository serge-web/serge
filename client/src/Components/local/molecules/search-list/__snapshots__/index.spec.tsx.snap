// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchList component: renders correctly 1`] = `
<div
  className="searchlist"
>
  <section
    className="input-container disable-offset fullwidth"
  >
    <div
      className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <div
        className="MuiInputBase-root MuiInput-root makeStyles-root-4 makeStyles-root-6 MuiInput-underline makeStyles-underline-5 makeStyles-underline-7 MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          name=""
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Select template"
          required={false}
          type="text"
          value=""
        />
      </div>
    </div>
  </section>
  <div
    className="searchlist-list searchlist-list"
  >
    <span
      className="searchlist-row active"
      onClick={[Function]}
    >
      <span>
        Label
      </span>
      <div
        className="searchlist-menu"
      />
    </span>
    <span
      className="searchlist-row active"
      onClick={[Function]}
    >
      <span>
        Label
      </span>
      <div
        className="searchlist-menu"
      />
    </span>
    <span
      className="searchlist-row active"
      onClick={[Function]}
    >
      <span>
        Label
      </span>
      <div
        className="searchlist-menu"
      />
    </span>
    <span
      className="searchlist-row active"
      onClick={[Function]}
    >
      <span>
        Label
      </span>
      <div
        className="searchlist-menu"
      />
    </span>
  </div>
</div>
`;
